File: flutter_inappwebview_android/android/src/main/java/com/pichillilorenzo/flutter_inappwebview_android/omid/OmidSessionManager.java
package com.pichillilorenzo.flutter_inappwebview_android.omid;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;
import android.text.TextUtils;

import androidx.annotation.NonNull;

import com.iab.omid.library.megabrainco.Omid;
import com.iab.omid.library.megabrainco.adsession.AdSession;
import com.iab.omid.library.megabrainco.adsession.AdSessionConfiguration;
import com.iab.omid.library.megabrainco.adsession.AdSessionContext;
import com.iab.omid.library.megabrainco.adsession.CreativeType;
import com.iab.omid.library.megabrainco.adsession.ImpressionType;
import com.iab.omid.library.megabrainco.adsession.Owner;
import com.iab.omid.library.megabrainco.adsession.Partner;
import com.pichillilorenzo.flutter_inappwebview_android.InAppWebViewFlutterPlugin;
import com.pichillilorenzo.flutter_inappwebview_android.types.ChannelDelegateImpl;
import com.pichillilorenzo.flutter_inappwebview_android.webview.WebViewInstanceRegistry;
import com.pichillilorenzo.flutter_inappwebview_android.webview.in_app_webview.InAppWebView;

import java.lang.ref.WeakReference;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

public class OmidSessionManager extends ChannelDelegateImpl {
  public static final String METHOD_CHANNEL_NAME = "kontext_omid";
  private static final int WEBVIEW_HOLD_DURATION_MS = 1100;

  @NonNull
  private final InAppWebViewFlutterPlugin plugin;
  @NonNull
  private final Handler mainHandler = new Handler(Looper.getMainLooper());
  @NonNull
  private final Map<String, OmidSessionState> sessions = new ConcurrentHashMap<>();

  public OmidSessionManager(@NonNull InAppWebViewFlutterPlugin plugin) {
    super(new MethodChannel(plugin.messenger, METHOD_CHANNEL_NAME));
    this.plugin = plugin;
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
    switch (call.method) {
      case "startOmidSession":
        startSession(call, result);
        break;
      case "stopOmidSession":
        stopSession(call, result);
        break;
      default:
        result.notImplemented();
        break;
    }
  }

  private void startSession(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
    final String instanceId = call.argument("instanceId");
    if (TextUtils.isEmpty(instanceId)) {
      result.error("missing_instance_id", "Expected a non-empty instanceId.", null);
      return;
    }
    if (sessions.containsKey(instanceId)) {
      result.error("duplicate_session", "An OMID session already exists for instanceId " + instanceId + ".", null);
      return;
    }

    final InAppWebView webView = WebViewInstanceRegistry.get(instanceId);
    if (webView == null) {
      result.error("webview_not_found", "No WebView found for instanceId " + instanceId + ".", null);
      return;
    }

    final String partnerName = call.argument("partnerName");
    final String partnerVersion = call.argument("partnerVersion");
    if (TextUtils.isEmpty(partnerName) || TextUtils.isEmpty(partnerVersion)) {
      result.error("invalid_partner", "partnerName and partnerVersion are required.", null);
      return;
    }

    final String contentUrl = call.argument("contentUrl");
    final String customReferenceData = call.argument("customReferenceData");

    if (contentUrl != null && contentUrl.length() > 512) {
      result.error("invalid_content_url", "contentUrl must be 512 characters or fewer.", null);
      return;
    }
    if (customReferenceData != null && customReferenceData.length() > 256) {
      result.error("invalid_custom_reference_data", "customReferenceData must be 256 characters or fewer.", null);
      return;
    }

    mainHandler.post(() -> {
      if (sessions.containsKey(instanceId)) {
        result.error("duplicate_session", "An OMID session already exists for instanceId " + instanceId + ".", null);
        return;
      }
      if (!ensureOmidActivated(result)) {
        return;
      }
      try {
        Partner partner = Partner.createPartner(partnerName, partnerVersion);
        AdSessionContext context = AdSessionContext.createHtmlAdSessionContext(
                partner,
                webView,
                contentUrl,
                customReferenceData
        );
        AdSessionConfiguration configuration = AdSessionConfiguration.createAdSessionConfiguration(
                CreativeType.HTML_DISPLAY,
                ImpressionType.BEGIN_TO_RENDER,
                Owner.JAVASCRIPT,
                Owner.NONE,
                false
        );
        AdSession adSession = AdSession.createAdSession(configuration, context);
        adSession.registerAdView(webView);
        adSession.start();

        sessions.put(instanceId, new OmidSessionState(webView, adSession));
        result.success(true);
      } catch (Exception e) {
        result.error("session_start_failed", "Failed to start OMID session: " + e.getMessage(), null);
        return;
      }
    });
  }

  private void stopSession(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
    final String instanceId = call.argument("instanceId");
    if (TextUtils.isEmpty(instanceId)) {
      result.error("missing_instance_id", "Expected a non-empty instanceId.", null);
      return;
    }
    mainHandler.post(() -> {
      if (!Omid.isActive()) {
        result.error("omid_inactive", "OM SDK is not active.", null);
        return;
      }
      OmidSessionState sessionState = sessions.remove(instanceId);
      if (sessionState == null) {
        result.error("session_not_found", "No active OMID session for instanceId " + instanceId + ".", null);
        return;
      }
      try {
        sessionState.adSession.finish();
      } catch (Exception e) {
        result.error("omid_stop_failed", "Failed to finish OMID session: " + e.getMessage(), null);
        return;
      }
      InAppWebView webView = sessionState.webViewRef.get();
      if (webView != null) {
        // The 1.1 second delay (OMID guidance) gives the system time to properly clean up
        // any remaining ad-related operations before allowing the WebView to be disposed of.
        webView.postDelayed(() -> webView.hashCode(), WEBVIEW_HOLD_DURATION_MS);
      }
      result.success(true);
    });
  }

  private boolean ensureOmidActivated(@NonNull MethodChannel.Result result) {
    if (Omid.isActive()) {
      return true;
    }
    Context context = plugin.applicationContext;
    if (context == null) {
      result.error("omid_activation_failed", "Application context is unavailable for OM SDK activation.", null);
      return false;
    }
    try {
      Omid.activate(context);
      return true;
    } catch (Exception e) {
      result.error("omid_activation_failed", "Failed to activate OM SDK: " + e.getMessage(), null);
      return false;
    }
  }

  private static class OmidSessionState {
    final WeakReference<InAppWebView> webViewRef;
    final AdSession adSession;

    OmidSessionState(@NonNull InAppWebView webView, @NonNull AdSession adSession) {
      this.webViewRef = new WeakReference<>(webView);
      this.adSession = adSession;
    }
  }
}

File: flutter_inappwebview_ios/ios/Classes/OmidSessionManager.swift
import Foundation
import WebKit
import OMSDK_Megabrainco

@available(iOS 11.0, *) public class OmidSessionManager: ChannelDelegate {
    static let methodChannelName = "kontext_omid"
    private var sessions: [String: OmidSessionState] = [:]

    private struct OmidSessionState {
        weak var webView: InAppWebView?
        let adSession: OMIDMegabraincoAdSession
    }

    init ? (plugin: SwiftFlutterPlugin) {
        guard let registrar = plugin.registrar else {
            return nil
        }
        let channel = FlutterMethodChannel(name: OmidSessionManager.methodChannelName, binaryMessenger: registrar.messenger())
        super.init(channel: channel)
    }

    public override func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?] else {
            result(FlutterError(code: "invalid_arguments", message: "Expected argument map.", details: nil))
            return
        }

        switch call.method {
        case "startOmidSession":
            startSession(arguments: arguments, result: result)
        case "stopOmidSession":
            stopSession(arguments: arguments, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    public func disposeSessions() {
        DispatchQueue.main.async {
            for state in self.sessions.values {
                state.adSession.finish()
            }
            self.sessions.removeAll()
        }
    }

    private func startSession(arguments: [String: Any?], result: @escaping FlutterResult) {
        guard let instanceId = arguments["instanceId"] as? String, !instanceId.isEmpty else {
            result(FlutterError(code: "missing_instance_id", message: "Expected a non-empty instanceId.", details: nil))
            return
        }
        guard sessions[instanceId] == nil else {
            result(FlutterError(code: "duplicate_session", message: "An OMID session already exists for instanceId \(instanceId).", details: nil))
            return
        }
        guard let webView = WebViewInstanceRegistry.get(instanceId: instanceId) else {
            result(FlutterError(code: "webview_not_found", message: "No WebView found for instanceId \(instanceId).", details: nil))
            return
        }
        guard let partnerName = arguments["partnerName"] as? String, !partnerName.isEmpty,
        let partnerVersion = arguments["partnerVersion"] as? String, !partnerVersion.isEmpty else {
            result(FlutterError(code: "invalid_partner", message: "partnerName and partnerVersion are required.", details: nil))
            return
        }
        if let customReferenceData = arguments["customReferenceData"] as? String, customReferenceData.count > 256 {
            result(FlutterError(code: "invalid_custom_reference_data", message: "customReferenceData must be 256 characters or fewer.", details: nil))
            return
        }

        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            guard self.sessions[instanceId] == nil else {
                result(FlutterError(code: "duplicate_session", message: "An OMID session already exists for instanceId \(instanceId).", details: nil))
                return
            }
            guard self.ensureOmidActivated(result: result) else {
                return
            }
            let partner = OMIDMegabraincoPartner(name: partnerName, versionString: partnerVersion)
            guard let partner = partner else {
                result(FlutterError(code: "session_start_failed", message: "Failed to create OMID partner instance.", details: nil))
                return
            }

            let contentUrl = arguments["contentUrl"] as? String
            let customReferenceData = arguments["customReferenceData"] as? String
            do {

                let adSessionContext = try OMIDMegabraincoAdSessionContext(
                    partner: partner,
                    webView: webView,
                    contentUrl: contentUrl,
                    customReferenceIdentifier: customReferenceData
                )

                let adSessionConfiguration = try OMIDMegabraincoAdSessionConfiguration(
                    creativeType: .htmlDisplay,
                    impressionType: .beginToRender,
                    impressionOwner: .javaScriptOwner,
                    mediaEventsOwner: .noneOwner,
                    isolateVerificationScripts: false
                )

                let adSession = try OMIDMegabraincoAdSession(
                    configuration: adSessionConfiguration,
                    adSessionContext: adSessionContext
                )

                adSession.mainAdView = webView
                adSession.start()
                self.sessions[instanceId] = OmidSessionState(webView: webView, adSession: adSession)
                result(true)
            } catch {
                result(FlutterError(code: "session_start_failed", message: "Failed to start OMID session.", details: error.localizedDescription))
            }
        }
    }

    private func stopSession(arguments: [String: Any?], result: @escaping FlutterResult) {
        guard let instanceId = arguments["instanceId"] as? String, !instanceId.isEmpty else {
            result(FlutterError(code: "missing_instance_id", message: "Expected a non-empty instanceId.", details: nil))
            return
        }
        DispatchQueue.main.async {
            if let state = self.sessions.removeValue(forKey: instanceId) {
                state.adSession.finish()
                result(true)
            } else {
                result(FlutterError(code: "session_not_found", message: "No OMID session for instanceId \(instanceId).", details: nil))
            }
        }
    }


    private func ensureOmidActivated(result: @escaping FlutterResult) -> Bool {
        let sdk = OMIDMegabraincoSDK.shared
        if sdk.isActive {
            return true
        }
        if sdk.activate() {
            return true
        }
        result(FlutterError(code: "omid_activation_failed", message: "Failed to activate OM SDK.", details: nil))
        return false
    }
}
